src/http/delta.rs

```diff
@@ -0,0 +1,327 @@
+use crate::payload::{
+    PayloadDelta, PayloadSnapshot, RouteOrigin, SharedHistory
+};
+use super::errors::{bad_request, initial_validation};
+
+//------------ handle_get ----------------------------------------------------
+
+pub fn handle_get(
+    req: &Request<Body>,
+    history: &SharedHistory,
+) -> Option<Response<Body>> {
+    if req.uri().path() != "/json-delta" {
+        return None
+    }
+    let history = history.read();
+
+    if !history.is_active() {
+        return Some(initial_validation())
+    }
+
+    let version = match version_from_query(req.uri().query()) {
+        Ok(version) => version,
+        Err(response) => return Some(response)
+    };
+
+    if let Some((session, serial)) = version {
+        if session == history.session() {
+            if let Some(delta) = history.delta_since(serial) {
+                return Some(handle_delta(
+                    session, serial, history.serial(), delta
+                ))
+            }
+        }
+    }
+
+    let snapshot = match history.current() {
+        Some(snapshot) => snapshot,
+        None => return Some(initial_validation()),
+    };
+    Some(handle_reset(history.session(), history.serial(), snapshot))
+}
+
+fn version_from_query(
+    query: Option<&str>
+) -> Result<Option<(u64, Serial)>, Response<Body>> {
+    let query = match query {
+        Some(query) => query,
+        None => return Ok(None)
+    };
+    let mut session = None;
+    let mut serial = None;
+
+    for (key, value) in form_urlencoded::parse(query.as_ref()) {
+        if key == "session" {
+            if session.is_some() {
+                return Err(bad_request());
+            }
+            session = Some(u64::from_str(&value).map_err(|_| bad_request())?);
+        }
+        else if key == "serial" {
+            if serial.is_some() {
+                return Err(bad_request());
+            }
+            serial = Some(Serial::from_str(&value).map_err(|_| bad_request())?);
+        }
+        else {
+            return Err(bad_request());
+        }
+    }
+    match (session, serial) {
+        (Some(session), Some(serial)) => Ok(Some((session, serial))),
+        (None, None) => Ok(None),
+        _ => Err(bad_request())
+    }
+}
+
+fn handle_delta(
+    session: u64, from_serial: Serial, to_serial: Serial,
+    delta: Arc<PayloadDelta>
+) -> Response<Body> {
+    Response::builder()
+    .header("Content-Type", "application/json")
+    .body(Body::wrap_stream(stream::iter(
+        DeltaStream::new(session, from_serial, to_serial, delta)
+        .map(Result::<_, Infallible>::Ok)
+    )))
+    .unwrap()
+}
+
+fn handle_reset(
+    session: u64, to_serial: Serial, snapshot: Arc<PayloadSnapshot>
+) -> Response<Body> {
+    Response::builder()
+    .header("Content-Type", "application/json")
+    .body(Body::wrap_stream(stream::iter(
+        SnapshotStream::new(session, to_serial, snapshot)
+        .map(Result::<_, Infallible>::Ok)
+    )))
+    .unwrap()
+}
+
```

Comment:
I have an issue somewhere for replacing those. It was initially part of 0.9, but I have punted it to 0.9.1 because I want to look at all the unwraps everywhere and replace those if at all possible.